src/ifcgeom/mapping/IfcLinearPlacement.cpp

```diff
@@ -0,0 +1,87 @@
+#define mapping POSTFIX_SCHEMA(mapping)
+using namespace ifcopenshell::geometry;
+
+#if defined SCHEMA_HAS_IfcLinearPlacement
+
+taxonomy::ptr mapping::map_impl(const IfcSchema::IfcLinearPlacement* inst) {
+
+	// The IFC specification does not provided a description of the optional
+	// CartesianPosition attribute. It is assumed to be a pre-computed IfcAxis2Placement3D
+	// to be used by software that don't support IfcLinearPlacement. In this case,
+	// we will simply use the intended CartesianPosition provided by the IFC model
+    if (inst->CartesianPosition()) {
+        return map(inst->CartesianPosition());
+    }
+
+	 // the following is taken from IfcLocalPlacement and tweaked a little
+	 // assumes that PlacementRelTo is relative to another IfcLinearPlacement
+	IfcSchema::IfcLinearPlacement* current = (IfcSchema::IfcLinearPlacement*)inst;
+	auto m4 = taxonomy::make<taxonomy::matrix4>();
+
+#if defined SCHEMA_HAS_IfcAxis2PlacementLinear
+	// IfcLinearPlacement was added in IFC 4.1 but it had an Orientation attribute of type IfcOrientationExpression, which when combined with other
+	// attributes was similar to IfcAxis2PlacementLinear. IFC 4.1 and IFC 4.2 have been withdrawn so I'm not going to try to implement linear placement for them.
+	// For this reason the preprocessor skips this code if SCHEMA_HAS_IfcAxix2PlacementLinear is not defined
+
+	for (;;) {
+        IfcSchema::IfcAxis2PlacementLinear* relplacement = current->RelativePlacement();
+        m4->components() = taxonomy::cast<taxonomy::matrix4>(map(relplacement))->ccomponents() * m4->ccomponents();
+
+		  if (current->PlacementRelTo()) {
+			IfcSchema::IfcObjectPlacement* parent = current->PlacementRelTo();
+
+			bool parent_placement_ignored = false;
+			if (placement_rel_to_type_ || placement_rel_to_instance_) {
+				IfcSchema::IfcProduct::list::ptr parent_places = parent->PlacesObject();
+				for (auto iter = parent_places->begin(); iter != parent_places->end(); ++iter) {
+					if ((placement_rel_to_type_ && (*iter)->declaration().is(*placement_rel_to_type_)) ||
+						(placement_rel_to_instance_ && (*iter)->as<IfcUtil::IfcBaseEntity>() == placement_rel_to_instance_)) {
+						parent_placement_ignored = true;
+					}
+				}
+			}
+
+			if (parent_placement_ignored) {
+				// The parent placement of the current is a placement for a type that is
+				// being ignored (Site or Building) or it is the host element of an opening.
+				break;
+            } else if (parent->declaration().is(IfcSchema::IfcLinearPlacement::Class())) {
+				// Keep processing parent placements
+                current = current->PlacementRelTo()->as<IfcSchema::IfcLinearPlacement>();
+			} else {
+				// This is the root placement (typically Site).
+				break;
+			}
+		} else {
+			break;
+		}
+	}
+#endif // SCHEMA_HAS_IfcAxis2PlacementLinear
+
+	// @todo: rb - not sure what this means... it came from IfcLocalPlacement
+	// m4->components() = offset_and_rotation_ * m4->components();
```

Comment:
@aothms Before I go too far down the wrong path, I think the implementations of IfcLocalPlacement.cpp and IfcLinearPlacement.cpp needs your critical review. Not knowing what to do, I copied IfcLocalPlacement.cpp and changed IfcLocalPlacement to IfcLinearPlacement.

Now that I understand the, for lack of knowing better terminology, the chain of RelativePlacement can be made up of any placement types, I don't think `for(;;)` loops are correct. The loop in IfcLocalPlacement expect the PlacementRelTo to be of type IfcLocalPlacement - which makes sence because there was only IfcLocalPlacement, but not any more. The loop in IfcLinearPlacement expects the PlacementRelTo to be of type IfcLinearPlacement.

I'm thinking IfcLocalPlacement.cpp and IfcLinearPlacement.cpp go away in favor of an IfcObjectPlacement.cpp that essentially works the same way, except that it can handle multiple IfcObjectPlacement subtypes. Maybe something like this, but I prefer it to be more polymorphic on IfcObjecPlacement.

```
#include "mapping.h"
#define mapping POSTFIX_SCHEMA(mapping)
using namespace ifcopenshell::geometry;

taxonomy::ptr mapping::map_impl(const IfcSchema::IfcObjectPlacement* inst) {
	IfcSchema::IfcObjectPlacement* current = inst;
	auto m4 = taxonomy::make<taxonomy::matrix4>();
	for (;;) {
		IfcSchema::IfcAxis2Placement* relplacement = current->RelativePlacement();
		// @todo this type check is wrong and unnecessary?
		if (relplacement->as<IfcSchema::IfcAxis2Placement3D>()) {
			// @todo check
			m4->components() = taxonomy::cast<taxonomy::matrix4>(map(relplacement))->ccomponents() * m4->ccomponents();
		}
		else if (relplacement->as<IfcSchema::IfcAxis2PlacementLinear>()) {
			// @todo check
			Matrix4d m_linear = taxonomy::cast<taxonomy::matrix4>(map(relplacement))->ccomponents();

			if (relplacement->CartesianPosition()) {
				auto m_fallback = taxonomy::cast<taxonomy::matrix4>(map(relplacement->CartesianPosition()));
				if (m_linear != m_fallback) {
					Logger::Warning("IfcLinearPlacement.CartesianPosition is different than the computed placement", inst);
				}
			}

			m4->components() = m_linear * m4->ccomponents();
		}

		if (current->PlacementRelTo()) {
			IfcSchema::IfcObjectPlacement* parent = current->PlacementRelTo();

			bool parent_placement_ignored = false;
			if (placement_rel_to_type_ || placement_rel_to_instance_) {
				IfcSchema::IfcProduct::list::ptr parent_places = parent->PlacesObject();
				for (auto iter = parent_places->begin(); iter != parent_places->end(); ++iter) {
					if ((placement_rel_to_type_ && (*iter)->declaration().is(*placement_rel_to_type_)) ||
						(placement_rel_to_instance_ && (*iter)->as<IfcUtil::IfcBaseEntity>() == placement_rel_to_instance_)) {
						parent_placement_ignored = true;
					}
				}
			}

			if (parent_placement_ignored) {
				// The parent placement of the current is a placement for a type that is
				// being ignored (Site or Building) or it is the host element of an opening.
				break;
			}
			else if (parent->declaration().is(IfcSchema::IfcLocalPlacement::Class())) {
				// Keep processing parent placements
				current = current->PlacementRelTo()->as<IfcSchema::IfcLocalPlacement>();
			}
			else if (parent->declaration().is(IfcSchema::IfcLinearPlacement::Class())) {
				// Keep processing parent placements
				current = current->PlacementRelTo()->as<IfcSchema::IfcLinearPlacement>();
			}
			else {
				// This is the root placement (typically Site).
				break;
			}
		}
		else {
			break;
		}
	}

	// @todo
	// m4->components() = offset_and_rotation_ * m4->components();

	return m4;
}
```

Reply:
Thank you, I missed that part. I'm thinking something else now though. Why do we unroll the recursion manually using this `for(;;)` loop instead of turning it simply into a recursive function.

```C++
taxonomy::ptr mapping::map_impl(const IfcSchema::IfcObjectPlacement* inst) {
    return taxonomy::make<taxonomy::matrix4>(
        taxonomy::cast<taxonomy::matrix4>(map(inst->RelativePlacement()))->ccomponents() *
        taxonomy::cast<taxonomy::matrix4>(map(inst->PlacementRelTo()))->ccomponents() 
    );
}
```

(this is oversimplified because IfcObjectPlacement does not have the RelativePlacement yet)

(it's even more efficient because we get caching for free along the chain)

Let me know what you think. If you prefer I can implement it like that after merging your PR.

Reply:
Not sure I completely understand your idea. Working at the level of IfcObjectPlacement makes since to me but the implementation also needs to deal with placements polymorphically so special cases can be addressed.

The one special case that comes to mind is that IfcLinearPlacement has the alternative CartesianPosition to deal with. I committed a change in IfcLinearPlacement (c8b99eed1119a5e21402cbe8670becead8bdec1a) (IfcLinearPlacement.cpp L76-L81) that compares the computed placement and the CartesianPosition and logs a warning if they are different. I think that behavior should remain. 

I'm deep into implementing IfcSegmentedGradientCurve at the moment, so if you have time to implement, that would be great. You also understand placements better that I. I think the PR is ready to be merged at any time (should I do the merge or you? - I don't feel like it is my repository to merge code into).